internal/oauth2/lazy_token.go

```diff
@@ -0,0 +1,105 @@
+package oauth2
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+// Tokener is an interface that defines the AccessToken method.
+type Tokener interface {
+	NextToken(context.Context) (string, error)
+	ForceRefresh(context.Context) (string, error)
+}
+
+type LazyToken struct {
+	// Url represents the URL used for acquiring the token.
+	Url string
+	// ClientId is the client ID used for authentication.
+	ClientId string
+	// ClientSecret is the client secret used for authentication.
+	ClientSecret string
+	// 	AccessToken string holds the currently cached token.
+	AccessToken string
+	// 	RefreshToken string token after refresh.
+	RefreshToken string
+	// Expiration stores the expiration time of the current token.
+	Expiration time.Time
+	// mutex ensures safe concurrent access to the token and expiration fields.
+	mutex sync.Mutex
+}
+
+type tokenResponse struct {
+	AccessToken string    `json:"access_token"`
+	Expiration  time.Time `json:"expiration"`
+}
```

Comment:
I think you are missing refresh token from here. AFAIK access token cannot be used to refresh token.

Reply:
But that doesn't match what the current code does... The current code never updates the refresh token...

Reply:
Yeah I realized that Keycloak instance of RH SSO is not configured to return refresh tokens at all, so disregard this. Let’s just acquire new token when it expires.